wolfcrypt/src/port/iotsafe/iotsafe.c

```diff
@@ -0,0 +1,1064 @@
+       const unsigned char* sig, unsigned int sigSz,
+       const unsigned char* hash, unsigned int hashSz,
+       const unsigned char* keyDer, unsigned int keySz,
+       int* result, void* ctx)
+{
+    byte sig_raw[IOTSAFE_ECC_KSIZE* 2];
+    int ret;
+    ecc_key key;
+    word32 r_size = IOTSAFE_ECC_KSIZE, s_size = IOTSAFE_ECC_KSIZE;
+    word32 inOutIdx = 0;
+    IOTSAFE *iot = &ssl->iotsafe;
+    byte pubkey_slot = iot->peer_cert_slot;
+
+    (void)ctx;
+
+    if (iot->enabled) {
+        /* Invalidate verification, by default. */
+        *result = 0;
+        WOLFSSL_MSG("IOTSAFE: Called wolfIoT_ecc_verify\n");
+
+        /* Convert ECC signature into R,S */
+        wc_ecc_sig_to_rs(sig, sigSz, sig_raw, &r_size, sig_raw
+                + IOTSAFE_ECC_KSIZE, &s_size);
+        /* Store the server's public key in IoT-safe vault */
+
+        /* Create an ecc object to handle the key */
+        ret = wc_ecc_init(&key);
+        if (ret != 0)
+            return ret;
+
+        /* Import from keyDer argument */
+        ret = wc_EccPublicKeyDecode(keyDer, &inOutIdx, &key, keySz);
+        if (ret < 0) {
+            WOLFSSL_MSG("Could not decode public key");
+            return ret;
+        }
+        /* Store public key in IoT-safe slot */
+        ret = iotsafe_put_public_key(pubkey_slot, &key);
+        if (ret < 0) {
+            WOLFSSL_MSG("Error storing public key");
+            return ret;
+        }
+        /* Call iotsafe_verify_hash with ECC256 + SHA256 */
+        ret = iotsafe_verify_hash(pubkey_slot,
+                IOTSAFE_HASH_SHA256, IOTSAFE_SIGN_ECDSA,
+                hash, hashSz, sig_raw, 2 * IOTSAFE_ECC_KSIZE,
+                result);
+    } else {
+        word32    idx = 0;
+        ecc_key   myKey;
+        ret = wc_ecc_init(&myKey);
+        if (ret == 0) {
+            ret = wc_EccPublicKeyDecode(keyDer, &idx, &myKey, keySz);
+            if (ret == 0)
+                ret = wc_ecc_verify_hash(sig, sigSz, hash, hashSz, result, &myKey);
+            wc_ecc_free(&myKey);
+        }
+    }
+    return ret;
+}
+
+static int wolfIoT_ecc_shared_secret(WOLFSSL* ssl, struct ecc_key* otherKey,
+        unsigned char* pubKeyDer, word32* pubKeySz,
+        unsigned char* out, word32* outlen,
+        int side, void* ctx)
+{
+    int ret;
+    char *resp;
+    ecc_key   tmpKey;
+    IOTSAFE *iotsafe = &ssl->iotsafe;
+    byte keypair_slot = iotsafe->ecdh_keypair_slot;
+    byte pubkey_idx = iotsafe->peer_pubkey_slot;
+
+    (void)ctx;
+
+    WOLFSSL_MSG("IOTSAFE: Called wolfIoT_ecc_shared_secret\n");
+
+    if (iotsafe->enabled) {
+        WOLFSSL_MSG("Generating ECDH key pair");
+        ret = iotsafe_gen_keypair(keypair_slot);
+        if (ret < 0) {
+            WOLFSSL_MSG("Error generating IoT-safe key pair");
+            return ret;
+        }
+
+        /* Importing generated public key */
+        ret = iotsafe_get_public_key(keypair_slot, &tmpKey);
+        if (ret < 0) {
+            return WC_HW_E;
```

Comment:
The tmpKey is not being free'd on errors... Should call `wc_ecc_free`.

Reply:
refactore with single return point after init to ensure that wc_ecc_free is called in all cases.